qcodes/instrument_drivers/AlazarTech/ATS.py

```diff
@@ -574,83 +614,146 @@ def acquire(self, mode=None, samples_per_record=None,
-        # buffer handling from acquisition
-        buffers_completed = 0
-        buffer_timeout = self.buffer_timeout._get_byte()
-        self.buffer_timeout._set_updated()
-
-        buffer_recycling = (self.buffers_per_acquisition._get_byte() >
-                            self.allocated_buffers._get_byte())
-
-        while buffers_completed < self.buffers_per_acquisition._get_byte():
-            buf = self.buffer_list[buffers_completed % allocated_buffers]
-
-            self._call_dll('AlazarWaitAsyncBufferComplete',
-                           self._handle, buf.addr, buffer_timeout)
-
-            # TODO(damazter) (C) last series of buffers must be handled
-            # exceptionally
-            # (and I want to test the difference) by changing buffer
-            # recycling for the last series of buffers
-
-            # if buffers must be recycled, extract data and repost them
-            # otherwise continue to next buffer
-
-            if buffer_recycling:
-                acquisition_controller.handle_buffer(buf.buffer)
+        logger.info("made buffer list length {}".format(len(self.buffer_list)))
+        try:
+            for buf in self.buffer_list:
+                self._ATS_dll.AlazarPostAsyncBuffer.argtypes = [ctypes.c_uint32, ctypes.c_void_p, ctypes.c_uint32]
                 self._call_dll('AlazarPostAsyncBuffer',
-                               self._handle, buf.addr, buf.size_bytes)
-            buffers_completed += 1
-
-        # stop measurement here
-        self._call_dll('AlazarAbortAsyncRead', self._handle)
-
+                               self._handle, ctypes.cast(buf.addr, ctypes.c_void_p), buf.size_bytes)
+            self.allocated_buffers._set_updated()
+
+            # -----start capture here-----
+            acquisition_controller.pre_start_capture()
+            start = time.clock()  # Keep track of when acquisition started
+            # call the startcapture method
+            self._call_dll('AlazarStartCapture', self._handle)
+            logger.info("Capturing %d buffers." % buffers_per_acquisition)
+
+            acquisition_controller.pre_acquire()
+
+            # buffer handling from acquisition
+            buffers_completed = 0
+            bytes_transferred = 0
+            buffer_timeout = self.buffer_timeout._get_byte()
+            self.buffer_timeout._set_updated()
+
+            buffer_recycling = (self.buffers_per_acquisition._get_byte() >
+                                self.allocated_buffers._get_byte())
+            done_setup = time.clock()
+            while (buffers_completed < self.buffers_per_acquisition._get_byte()):
+                # Wait for the buffer at the head of the list of available
+                # buffers to be filled by the board.
+                buf = self.buffer_list[buffers_completed % allocated_buffers]
+                self._ATS_dll.AlazarWaitAsyncBufferComplete.argtypes = [ctypes.c_uint32, ctypes.c_void_p, ctypes.c_uint32]
```

Comment:
do you need to update the types for every buffer you acquire. Couldn't this line be called only once at the beginning of the driver?
we want to keep this part of the driver as fast as possible